File: packages/google_maps_flutter/google_maps_flutter_android/android/src/main/java/io/flutter/plugins/googlemaps/HeatmapController.java
// Copyright 2013 The Flutter Authors. All rights reserved.
// Use of this source code is governed by a BSD-style license that can be
// found in the LICENSE file.

package io.flutter.plugins.googlemaps;

import androidx.annotation.NonNull;
import com.google.android.gms.maps.model.TileOverlay;
import com.google.maps.android.heatmaps.Gradient;
import com.google.maps.android.heatmaps.HeatmapTileProvider;
import com.google.maps.android.heatmaps.WeightedLatLng;
import java.util.List;

/** Controller of a single Heatmap on the map. */
public class HeatmapController implements HeatmapOptionsSink {
  private final @NonNull HeatmapTileProvider heatmap;
  private final @NonNull TileOverlay heatmapTileOverlay;

  /** Construct a HeatmapController with the given heatmap and heatmapTileOverlay. */
  HeatmapController(@NonNull HeatmapTileProvider heatmap, @NonNull TileOverlay heatmapTileOverlay) {
    this.heatmap = heatmap;
    this.heatmapTileOverlay = heatmapTileOverlay;
  }

  /** Remove the heatmap from the map. */
  void remove() {
    heatmapTileOverlay.remove();
  }

  /** Clear the tile cache of the heatmap in order to update the heatmap. */
  void clearTileCache() {
    heatmapTileOverlay.clearTileCache();
  }

  @Override
  public void setWeightedData(@NonNull List<WeightedLatLng> weightedData) {
    heatmap.setWeightedData(weightedData);
  }

  @Override
  public void setGradient(@NonNull Gradient gradient) {
    heatmap.setGradient(gradient);
  }

  @Override
  public void setMaxIntensity(double maxIntensity) {
    heatmap.setMaxIntensity(maxIntensity);
  }

  @Override
  public void setOpacity(double opacity) {
    heatmap.setOpacity(opacity);
  }

  @Override
  public void setRadius(int radius) {
    heatmap.setRadius(radius);
  }
}

File: packages/google_maps_flutter/google_maps_flutter_ios_sdk9/ios/google_maps_flutter_ios_sdk9/Sources/google_maps_flutter_ios_sdk9/HeatmapController.swift
// Copyright 2013 The Flutter Authors
// Use of this source code is governed by a BSD-style license that can be
// found in the LICENSE file.

import GoogleMaps
import GoogleMapsUtils

/// Controller of a single Heatmap on the map.
class HeatmapController {
  let heatmapTileLayer: GMUHeatmapTileLayer
  private weak var mapView: GMSMapView?

  init(heatmap: PlatformHeatmap, tileLayer: GMUHeatmapTileLayer, mapView: GMSMapView) {
    self.heatmapTileLayer = tileLayer
    self.mapView = mapView
    HeatmapController.update(tileLayer, from: heatmap, mapView: mapView)
  }

  func removeHeatmap() {
    heatmapTileLayer.map = nil
  }

  func clearTileCache() {
    heatmapTileLayer.clearTileCache()
  }

  func update(from platformHeatmap: PlatformHeatmap) {
    if let mapView = mapView {
      HeatmapController.update(heatmapTileLayer, from: platformHeatmap, mapView: mapView)
    }
  }

  /// Updates the underlying GMUHeatmapTileLayer with the properties from the given platform heatmap.
  ///
  /// Setting the heatmap to visible will set its map to the given mapView.
  static func update(
    _ heatmapTileLayer: GMUHeatmapTileLayer,
    from platformHeatmap: PlatformHeatmap,
    mapView: GMSMapView
  ) {
    heatmapTileLayer.weightedData = platformHeatmap.data.map({ $0.toGMUWeightedLatLng() })
    if let gradient = platformHeatmap.gradient {
      heatmapTileLayer.gradient = gradient.toGMUGradient()
    }
    heatmapTileLayer.opacity = Float(platformHeatmap.opacity)
    heatmapTileLayer.radius = UInt(platformHeatmap.radius)
    heatmapTileLayer.minimumZoomIntensity = UInt(platformHeatmap.minimumZoomIntensity)
    heatmapTileLayer.maximumZoomIntensity = UInt(platformHeatmap.maximumZoomIntensity)

    // The map must be set each time for options to update.
    // This must be done last, to avoid visual flickers of default property values.
    heatmapTileLayer.map = mapView
  }
}

/// Controller of multiple Heatmaps on the map.
class HeatmapsController {
  private var heatmapIdToController: [String: HeatmapController] = [:]
  private weak var mapView: GMSMapView?

  init(mapView: GMSMapView) {
    self.mapView = mapView
  }

  func add(_ heatmapsToAdd: [PlatformHeatmap]) {
    guard let mapView = mapView else { return }
    for heatmap in heatmapsToAdd {
      let heatmapTileLayer = GMUHeatmapTileLayer()
      let controller = HeatmapController(
        heatmap: heatmap,
        tileLayer: heatmapTileLayer,
        mapView: mapView
      )
      heatmapIdToController[heatmap.heatmapId] = controller
    }
  }

  func change(_ heatmapsToChange: [PlatformHeatmap]) {
    for heatmap in heatmapsToChange {
      if let controller = heatmapIdToController[heatmap.heatmapId] {
        controller.update(from: heatmap)
        controller.clearTileCache()
      }
    }
  }

  func removeHeatmaps(withIdentifiers identifiers: [String]) {
    for heatmapId in identifiers {
      if let controller = heatmapIdToController[heatmapId] {
        controller.removeHeatmap()
        heatmapIdToController.removeValue(forKey: heatmapId)
      }
    }
  }

  func hasHeatmap(withIdentifier identifier: String) -> Bool {
    return heatmapIdToController[identifier] != nil
  }

  func heatmap(withIdentifier identifier: String) -> PlatformHeatmap? {
    guard let controller = heatmapIdToController[identifier] else { return nil }
    let heatmap = controller.heatmapTileLayer
    return PlatformHeatmap(
      heatmapId: identifier,
      data: heatmap.weightedData.map { PlatformWeightedLatLng.make(from: $0) },
      gradient: PlatformHeatmapGradient.make(from: heatmap.gradient),
      opacity: Double(heatmap.opacity),
      radius: Int64(heatmap.radius),
      minimumZoomIntensity: Int64(heatmap.minimumZoomIntensity),
      maximumZoomIntensity: Int64(heatmap.maximumZoomIntensity)
    )
  }
}
